File: 0067-add-binary/0067-add-binary.c
#define MAX(first, second) (( first ) > ( second ) ? ( first ) : ( second ))

#define SWAP(first, second, type) {     \
            type temp = *( first );     \
            *( first ) = *( second );   \
            *( second ) = temp;         \
        }

void reverse(char* str, size_t size) {
    char* left = str;
    char* right = str + size - 1;

    while (left != right && left != right + 1) {
        SWAP(left, right, char);
        ++left;
        --right;
    }
}

char* addBinary(char* a, char* b) {
    size_t aSize = strlen(a);
    size_t bSize = strlen(b);
    size_t resMaxSize = MAX(aSize, bSize) + 2;

    char* res = (char*)malloc(sizeof(char) * resMaxSize);

    if (res == NULL) {
        return NULL;
    }

    size_t aCurr = aSize;
    size_t bCurr = bSize;
    size_t resCurr = 0;
    unsigned char carry = 0;

    while (aCurr > 0 && bCurr > 0) {
        unsigned char aBit = a[--aCurr] - '0';
        unsigned char bBit = b[--bCurr] - '0';

        res[resCurr++] = aBit ^ bBit ^ carry + '0';
        carry = (aBit & bBit) | (aBit & carry) | (bBit & carry);
    }

    while (aCurr > 0) {
        unsigned char aBit = a[--aCurr] - '0';
        res[resCurr++] = aBit ^ carry + '0';
        carry = aBit & carry;
    }
 
    while (bCurr > 0) {
        unsigned char bBit = b[--bCurr] - '0';
        res[resCurr++] = bBit ^ carry + '0';
        carry = bBit & carry;
    }

    if (carry > 0) {
        res[resCurr++] = '1';
    }

    res[resCurr] = '\0';

    reverse(res, resCurr);
    return res;
}

File: 0067-add-binary/0067-add-binary.cs
public class Solution {
    public string AddBinary(string a, string b) {
        if (b.Length > a.Length) {
            return AddBinary(b, a);
        }

        char[] resArr = new char[a.Length + 1];
        resArr[a.Length] = '\0';

        int carry = 0;
        int bI = b.Length - 1;
        int aI = a.Length - 1;
        int resI = 0;

        while (aI >= 0 && bI >= 0) {
            int aBit = a[aI--] - '0';
            int bBit = b[bI--] - '0';

            int resBit = aBit ^ bBit ^ carry;
            carry = (aBit & bBit) | (aBit & carry) | (bBit & carry);

            resArr[resI++] = (char)('0' + resBit);
        }

        void handleRemaining(string x, ref int xI) {
            while (xI >= 0) {
                int xBit = x[xI--] - '0';

                int resBit = xBit ^ carry;
                carry = xBit & carry;

                resArr[resI++] = (char)('0' + resBit);
            }
        }

        handleRemaining(a, ref aI);
        handleRemaining(b, ref bI);

        if (carry > 0) {
            resArr[resI] = '1';
        } else {
            Array.Resize(ref resArr, resArr.Length - 1);
        }

        Array.Reverse(resArr);
        return new string(resArr);
    }
}